### DISCUSSION TITLE
Multiple topics callbacks on single subscriber [14483]

### DISCUSSION THREAD

**Discussion by paulbkeelp** [eProsima/Fast-DDS#2663] [Support]
We have a central subscriber application that need to subscribe 10+ Topics

For single Topic subscription all pretty much easy, we can call Init() all the default stuff and call the below callbacks
SubListener::on_subscription_matched
SubListener::on_data_available

If we have an application that need to subscribe 10+ Topics. How to acheive this easily - any examples please ?
SubListener::on_subscription_matched
SubListener::on_data_available

how to listen on multiple callbacks (multiple topics) on on_data_available. I see on MQTT we can register the callback on a topics. are there a simililar way please ?

**Comment by JLBuenoLopez** [comment]
Hi @paulbkeelp,

Using DDS you will need to create a specific [DataReader](https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/subscriber/dataReader/dataReader.html#datareader) for each Topic, as [`create_datareader`](https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/subscriber/dataReader/createDataReader.html#creating-a-datareader) requires the Topic as mandatory argument. Therefore, each DataReader is bound to a single Topic since its creation. These DataReaders can be created within the same [Subscriber](https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/subscriber/subscriber/subscriber.html#subscriber).

In order to be notified of data reception, you can use either the `on_data_on_readers` callback defined by the [`SubscriberListener`](https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/subscriber/subscriberListener/subscriberListener.html#subscriberlistener) or attach the same [DataReaderListener](https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/subscriber/dataReaderListener/dataReaderListener.html#datareaderlistener) to every DataReader.

I hope this information helps with your use case!

**Reply by paulbkeelp** [reply]
@JLBuenoLopez-eProsima Glad for your input. I will try that and Keep you updated.
Many Thanks

**Reply by zixuan-su** [reply]
hello, I'm having the same problem，do you solve it?

**Reply by EduPonz** [reply]
Hi @zixuan-su ,

To which problem are you referring? In DDS, each DataReader receives data from 1 topic, so you get callbacks each time any DataWriter publishes data on that topic. If you want to listen for data over different topics, you need several DataReaders (on per topic). What @JLBuenoLopez-eProsima is suggesting is a convenience to get a notification whenever any DataReader receives data, so that application can call to `read/take` on the appropriate one.

**Reply by zixuan-su** [reply]
> Hi @zixuan-su ,
> 
> To which problem are you referring? In DDS, each DataReader receives data from 1 topic, so you get callbacks each time any DataWriter publishes data on that topic. If you want to listen for data over different topics, you need several DataReaders (on per topic). What @JLBuenoLopez-eProsima is suggesting is a convenience to get a notification whenever any DataReader receives data, so that application can call to `read/take` on the appropriate one.

thanks， I have solved it!
